src/main/java/com/marklogic/spark/reader/optic/OpticReadContext.java

```diff
@@ -130,20 +172,55 @@ private PlanBuilder.Plan buildPlanForBucket(RowManager rowManager, PlanAnalysis.
             .bindParam("ML_LOWER_BOUND", bucket.lowerBound)
             .bindParam("ML_UPPER_BOUND", bucket.upperBound);
 
+        // Add user-defined parameters from options
+        for(Map.Entry<String, String> entry : this.userDefinedParams.entrySet()) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Adding user param: {}", entry);
+            }
+            plan = plan.bindParam(entry.getKey(), entry.getValue());
+        }
+
         if (opticFilters != null) {
             for (OpticFilter opticFilter : opticFilters) {
                 plan = opticFilter.bindFilterValue(plan);
             }
         }
 
+        for (Map.Entry<String, String> entry : columnParamValues.entrySet()) {
+            if (entry.getValue() == null) {
+                entry.setValue(columnParamDefaults.get(entry.getKey()));
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Setting default param column: {}", entry);
+                }
+            } else if (logger.isDebugEnabled()) {
+                logger.debug("Setting param column: {}", entry);
+            }
+
+            plan = plan.bindParam(entry.getKey(), entry.getValue());
+        }
+
         return plan;
     }
 
     void pushDownFiltersIntoOpticQuery(List<OpticFilter> opticFilters) {
         this.opticFilters = opticFilters;
         // Add each filter in a separate "where" so we don't toss an op.sqlCondition into an op.and,
         // which Optic does not allow.
-        opticFilters.forEach(filter -> addOperatorToPlan(PlanUtil.buildWhere(filter)));
+        opticFilters.forEach(filter -> {
```

Comment:
This should be another private method (lots of small private methods with self-documenting names = clean code, easy to maintain and read. Long methods = bad). 

This seems like the real key logic - i.e. check each filter in the Spark program to see if it corresponds to a user param name, and if so, add the filter value to the map that contains user params with values defined at runtime. Right?